File: 20.valid-parentheses.cpp
/*
 * @lc app=leetcode id=20 lang=cpp
 *
 * [20] Valid Parentheses
 */

#include <assert.h>

#include <stack>
#include <string>

using namespace std;

// @lc code=start
class Solution {
 public:
  bool isValid(string s) {
    int len = s.length();
    stack<char> vec;
    for (int i = 0; i < len; i++) {
      switch (s[i]) {
        case '(':
        case '{':
        case '[':
          vec.push(s[i]);
          break;
        case ']':
          if (!vec.size()) return false;
          if (vec.top() != '[') return false;
          vec.pop();
          break;
        case '}':
          if (!vec.size()) return false;
          if (vec.top() != '{') return false;
          vec.pop();
          break;
        case ')':
          if (!vec.size()) return false;
          if (vec.top() != '(') return false;
          vec.pop();
          break;
        default:
          break;
      }
    }

    if (vec.size()) return false;
    return true;
  }
};
// @lc code=end

int main() {
  Solution s;
  assert(!s.isValid("("));
  assert(!s.isValid("]"));
  assert(s.isValid("()"));
  assert(s.isValid("()[]{}"));
  assert(!s.isValid("(]"));
  assert(!s.isValid("([)]"));
  assert(s.isValid("{[]}"));
}

File: 20.valid-parentheses.rs
/*
 * @lc app=leetcode id=20 lang=rust
 *
 * [20] Valid Parentheses
 */

struct Solution;

// @lc code=start
impl Solution {
    pub fn is_valid(s: String) -> bool {
        let mut stack = vec![];
        for c in s.chars() {
            if c == '(' || c == '{' || c == '[' {
                stack.push(c);
                continue;
            }

            match stack.pop() {
                Some(last) => {
                    if (last == '(' && c == ')')
                        || (last == '{' && c == '}')
                        || (last == '[' && c == ']')
                    {
                        continue;
                    }
                    return false;
                }
                None => return false,
            }
        }
        stack.is_empty()
    }
}
// @lc code=end

#[test]
fn test1() {
    assert!(Solution::is_valid("()".to_string()));
    assert!(Solution::is_valid("()[]{}".to_string()));
    assert!(Solution::is_valid("{[]}".to_string()));
}

#[test]
fn test2() {
    assert!(!Solution::is_valid("(]".to_string()));
    assert!(!Solution::is_valid("([)]".to_string()));
}
